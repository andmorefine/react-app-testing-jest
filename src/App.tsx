import React, { useState, useEffect, useRef } from 'react'
import { getFormattedUserName } from './utility'
import './App.css'

// ユーザーデータの型を定義
interface User {
  id: number
  name: string
  username: string
}

const App = () => {
  const [users, setUsers] = useState<User[]>([])
  const isMounted = useRef(true) // isMountedをuseRefで管理

  // Fetch the data from the server
  useEffect(() => {
    const url: string = 'https://jsonplaceholder.typicode.com/users'
    const getUsers = async () => {
      const response = await fetch(url)
      if (!response.ok) {
        console.error('Failed to fetch users')
        return
      }
      const data: User[] = await response.json()
      setUsers(data)
    }
    getUsers()
    // Cleanup function to handle component unmounting
    return () => {
      isMounted.current = false // アンマウント時には.currentをfalseにする
    }
  }, [])

  return (
    <div>
      <h1>Users:</h1>
      <ul>
        {users.map(({ id, name, username }) => (
          <li key={id}>
            {name} -- <span>({getFormattedUserName(username)})</span>
          </li>
        ))}
      </ul>
    </div>
  )
}

export default App
