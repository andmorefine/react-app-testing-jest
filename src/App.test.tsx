import React from 'react'
import renderer, { act } from 'react-test-renderer'
import App from './App'

// ユニットテストで使用するダミーのユーザーデータ
const fakeUserData = [{ id: 1, name: 'Yamada Hanako', username: 'Yamada' }]

beforeEach(() => {
  global.fetch = jest.fn(() =>
    Promise.resolve(
      new Response(JSON.stringify(fakeUserData), {
        status: 200,
        headers: { 'Content-type': 'application/json' },
      }),
    ),
  )
})

afterEach(() => {
  jest.restoreAllMocks()
})

test('check if it renders a correct snapshot', async () => {
  let tree = renderer.create(<App />)
  await act(async () => {
    tree.update(<App />)
  })
  expect(tree.toJSON()).toMatchSnapshot()
})
