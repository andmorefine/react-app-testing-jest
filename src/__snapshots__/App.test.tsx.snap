// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check if it renders a correct snapshot 1`] = `
<div>
  <h1>
    Users:
  </h1>
  <ul>
    <li>
      Yamada Hanako
       -- 
      <span>
        (
        @Yamada
        )
      </span>
    </li>
  </ul>
</div>
`;
